src/uct/ib/base/ib_iface.c

```diff
@@ -743,23 +750,29 @@ static ucs_status_t uct_ib_iface_init_pkey(uct_ib_iface_t *iface,
             continue;
         }
 
-        /* take only the lower 15 bits for the comparison */
-        if ((config->pkey == UCS_HEXUNITS_AUTO) ||
-            ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey)) {
-            if (!(pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) &&
-                (/* limited PKEY hasn't been find yet */
-                 (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY) ||
-                 /* find more suitable PKEY with limited membership */
-                 ((pkey & UCT_IB_PKEY_PARTITION_MASK) == config->pkey))) {
-                lim_pkey_index = pkey_index;
-                lim_pkey       = pkey;
-                continue;
+        if (/* PKEY with full membership */
+            pkey & UCT_IB_PKEY_MEMBERSHIP_MASK) {
+            if (/* a specific PKEY was not requested */
+                (config->pkey == UCS_HEXUNITS_AUTO) ||
+                /* matches the requested PKEY */
+                uct_ib_iface_config_check_pkey(config, pkey)) {
+                /* found PKEY with full membership that can be used, stop
+                 * searching */
+                iface->pkey_index = pkey_index;
+                iface->pkey       = pkey;
+                pkey_found        = 1;
+                break;
             }
-
-            iface->pkey_index = pkey_index;
-            iface->pkey       = pkey;
-            pkey_found        = 1;
-            break;
+        } else if (/* a specific PKEY was not requested and no
+                    * PKEY with limited membership was choosen */
+                   ((config->pkey == UCS_HEXUNITS_AUTO) &&
+                    (lim_pkey == UCT_IB_ADDRESS_INVALID_PKEY)) ||
+                   /* matches the requested PKEY */
+                   (uct_ib_iface_config_check_pkey(config, pkey))) {
+            /* save PKEY with limited membership for futher use if more suitable
+             * PKEY is not found */
+            lim_pkey_index = pkey_index;
+            lim_pkey       = pkey;
```

Comment:
this looks even more complicated..

Reply:
thanks for pointing me out at this code, now it fixed to be simple and w/o additional `if`s that were not needed